client_code/kompot/_persist.py

```diff
@@ -0,0 +1,279 @@
+# SPDX-License-Identifier: MIT
+# Copyright (c) 2021 anvilistas
+
+"""A module to provide persistence mechanisms for a class.
+
+e.g.
+
+For a class named 'Book', we might have a data table named 'book' where each row
+corresponds to instance of the Book class. We might also have server functions to
+add a new row or to update existing entries:
+
+@row_backed_class
+class Book:
+    pass
+
+Create a new Book instance and fetch its row from the db:
+
+book = Book()
+book.get(title="Fluent Python")
+
+This assumes a server function exists named 'get_book', which take a title argument
+and returns a data tables row.
+
+All the columns from the row will now be available as attributes on the Book instance.
+e.g.
+
+print(book.title)
+>>> Fluent Python
+"""
+
+import anvil.server
+
+from ._register import register
+from ._rpc import call
+from ._serialize import serialize
+
+__version__ = "0.0.1"
+
+
+@register
+@anvil.server.portable_class
+class Dispatcher:
+    """A portable class for sending payloads from client to server
+
+    When changes are made to a persisted object, a Dispatcher instance records those
+    changes so they can be sent to the server if required.
+
+    This is particularly useful for classes backed by data tables rows since a row
+    cannot be created or updated on the client.
+    """
+
+    def __init__(self):
+        self._changed_attrs = set()
+
+    def __setattr__(self, name, value):
+        if name.startswith("_"):
+            object.__setattr__(self, name, value)
+        else:
+            self._changed_attrs.add(name)
+            object.__setattr__(self, name, value)
+
+    @property
+    def synced(self):
+        """Whether there are any pending changes recorded."""
+        return not self._changed_attrs
+
+    def clear(self):
+        """Clear any pending changes."""
+        for attr in self._changed_attrs:
+            delattr(self, attr)
+        self._changed_attrs = set()
+
+
+class LinkedAttribute:
```

Comment:
Oh linked attribute updates - fun

Does this require the writer to know which name they've mapped to which linked attribute,
and then handle this in reverse when doing server row update?

I'd be curious to see an example of this. It feels like there should be a cleaner way 🤷

Reply:
I agree. I wonder if set_name might help her somehow

Reply:
At the moment, the API for these looks something like:

```py
@row_backed_class
class Book:
    author_name = LinkedAttribute(name="author_name", linked_table="author", linked_attr="name")
```

assuming a `book` table with an `author` linked column and a linked table with a `name` column.